core/src/main/java/org/apache/accumulo/fate/zookeeper/ZooCacheFactory.java

```diff
@@ -27,6 +29,53 @@
 public class ZooCacheFactory {
   // TODO: make this better - LRU, soft references, ...
   private static Map<String,ZooCache> instances = new HashMap<>();
+  private static boolean enabled = true;
+
+  private static boolean isEnabled() {
+    synchronized (instances) {
+      return enabled;
+    }
+  }
+
+  private static void enable() {
+    synchronized (instances) {
+      enabled = true;
+    }
+  }
+
+  private static void disable() {
+    synchronized (instances) {
+      try {
+        instances.values().forEach(zc -> zc.close());
+      } finally {
+        instances.clear();
+        enabled = false;
+      }
+    }
+  }
+
+  static {
+    // important because of ZOOKEEPER-2368.. when zookeeper client is closed it does not generate an
+    // event!
+    SingletonManager.register(new SingletonService() {
+
+      @Override
+      public synchronized boolean isEnabled() {
+        return ZooCacheFactory.isEnabled();
+      }
+
+      @Override
+      public synchronized void enable() {
+        ZooCacheFactory.enable();
+      }
+
+      @Override
+      public synchronized void disable() {
+        ZooCacheFactory.disable();
+      }
+    });
+
+  }
```

Comment:
Could you register ZooCache only for clients?  I am wondering if this could hurt performance with locking on the server side.

Reply:
Its possible that it could cause server side performance problems. I would like to avoid complicating this code w/o evidence of that.  Also, if #735 is done this code may be significantly refactored.